source/common/modules/markdown-editor/parser/citation-parser.ts

```diff
@@ -8,51 +8,729 @@
+        const citekeyEnd = i
+
+        // In this branch, we only temporarliy collect all elements, since we do
+        // not yet know if the bracket actually closes. If it doesn't, anything
+        // until the citekey is still valid, but the rest must be thrown away.
+        const temporaryParts: MDElement[] = []
+
+        i++
+        temporaryParts.push(ctx.elt(NODES.MARK, i, ++i))
+        let intextSuffixStart = i
+
+        // Does the remaining slice start with an explicit locator label?
+        let locatorStart = -1
+        const slice = ctx.slice(i, Math.max(i + maxLocatorLabelLength, ctxEndPos)).toLowerCase()
+        // NOTE that we require each label to be followed by a space
+        const explicitLabel = allValidLocatorLabels.find(x => slice.startsWith(x + ' '))
+        if (explicitLabel !== undefined) {
+          // First, check if there are only punctuation marks and spaces between
+          // the citekey end and the locator start. If not, we should not detect
+          // this as a locator.
+          if (/^[\s,\.:;+-]*$/.test(ctx.slice(intextSuffixStart, i - 1))) {
+            // Found a valid locator label -> begin explicit locator
+            locatorStart = i
+            // Move i forward until after the space so that the implicit locator
+            // logic can take over
+            i += explicitLabel.length + 1
+          }
+        } else if (((ctx.char(i) >= 48 && ctx.char(i) <= 57) || ROMAN_NUMERAL_CODES.includes(ctx.char(i)))) {
+          // First, check if there are only punctuation marks and spaces between
+          // the citekey end and the locator start. If not, we should not detect
+          // this as a locator.
+          if (/^[\s,\.:;+-]*$/.test(ctx.slice(intextSuffixStart, i - 1))) {
+            // Found a valid locator character -> begin implicit locator
+            locatorStart = i
+          }
+        }
+
+        if (locatorStart > -1) {
+          // There was an implicit or explicit locator; so now we just have to
+          // move i forward until no more valid locator chars exist
+          while (i < ctxEndPos && ((ctx.char(i) >= 48 && ctx.char(i) <= 57) || ROMAN_NUMERAL_CODES.includes(ctx.char(i)) || ctx.char(i) === CHAR.HYPHEN)) {
+            i++
+          }
+
+          temporaryParts.push(ctx.elt(NODES.LOCATOR, locatorStart, i))
+          intextSuffixStart = i
+        } // Else: No locator, so essentially everything is suffix.
+
+        // Finally, we just have to find the closing bracket to complete the
+        // inline suffix.
+        while (i < ctxEndPos && ctx.char(i) !== CHAR.BRACKET_CLOSE) {
+          i++
+        }
+
+        if (ctx.char(i) === CHAR.BRACKET_CLOSE) {
+          // First, commit the temporary collected parts ...
+          parts.push(...temporaryParts)
+          // ... add the suffix (intextSuffixStart is sensitive to locator) ...
+          parts.push(ctx.elt(NODES.SUFFIX, intextSuffixStart, i))
```

Comment:
Here there's a check missing if locatorEnd === i, because otherwise we're mounting an empty node.